frontend/src/utils/userContext.js

```diff
@@ -0,0 +1,56 @@
+import { createContext, useState, useEffect } from "react";
+import { loggedInUser } from "../services/userService";
+import { useLocation, useNavigate } from "react-router-dom";
+
+import { logout } from "../services/userService";
+
+const UserContext = createContext(null);
+
+const UserContextProvider = ({ children }) => {
+  const [user, setUser] = useState(null);
+  const [loading, setLoading] = useState(true);
+  const location = useLocation();
+  const navigate = useNavigate();
+
+  const logIn = (data) => {
+    setUser(data);
+  };
+  const logOut = () => {
+    setUser(null);
+    logout();
+    navigate("/login");
+  };
+
+  useEffect(() => {
+    const fetchUser = () => {
+      loggedInUser()
+        .then((user) => {
+          setUser(user.data);
+          setLoading(false);
+        })
+        .catch((error) => {
+          console.log("An error occured");
+          if (
+            location.pathname === "/home" ||
+            location.pathname === "/profile" ||
+            location.pathname === "/profile/follows" ||
+            location.pathname === "/profile/edit"
+          ) {
+            navigate("login");
+          }
```

Comment:
Create a consts file for this because overtime it will go bigger and this context will be messy

Under consts folder, create routes.js then make list of routes there example:

const guestRoutes = [
   '/login',
  ....
];

cont authenticatedRoutes = [
  '/home',
   .....
]

then use it like this
```suggestion
          
      authenticatedRoutes.includes(location.pathname) && navigate("login");
```